File: 2021/day21/go/day21.go
package main

import "fmt"

// rolls the dice 3 times and returns the result and new dice state
func roll_3_times(rolls *int) int {
	move := 0
	for i := 0; i < 3; i++ {
		move += *rolls%100 + 1
		*rolls++
	}
	return move
}

func main() {
	positions := [2]int{6, 8}
	scores := [2]int{0, 0}
	rolls := 0
	for {
		for player := range positions {
			move := roll_3_times(&rolls)
			positions[player] = ((positions[player] + move) % 10)
			scores[player] += positions[player] + 1
			if scores[player] >= 1000 {
				fmt.Println(rolls * scores[1-player])
				return
			}
		}
	}
}

File: 2021/day21/nim/day21.nim
var turn=0
var players = [6,8] 
var scores = [0,0]
var rolls=0
var dice=0

while true:
    rolls+=3
    var n = -1
    for i in countup(1,3):
        dice+=1
        if dice==101:
            dice=1
        n+=dice
    players[turn]=1+(players[turn]+n) mod 10
    scores[turn]+=players[turn]
    if scores[turn]>=1000:
        break
    turn=1-turn

echo min(scores)*rolls
